### ISSUE TITLE
reg-tests/server/cli_set_ssl.vtc instability on aarch64

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#1547]
### Detailed Description of the Problem

reg-tests/server/cli_set_ssl.vtc  [fails](https://app.travis-ci.com/github/haproxy/haproxy/jobs/559445294) on arm64-graviton2

personally, I do not have graviton2, but I could reproduce failure on  [Ampere A1](https://www.oracle.com/cloud/compute/arm/)  in [Oracle Cloud](https://www.oracle.com/cloud/free/).

instability seems to be CPU dependent (the more cores, the higher failure rate). test was executed 1000 times

| OCPU count| number of failures (per 1000 execution) |
|--|--|
| 1| 0 |
| 2| 3 |
| 4| 33 |

mentioned test is not the only one unstable, other unstable tests are:

reg-tests/http-errorfiles/http_errors.vtc
reg-tests/server/cli_set_ssl.vtc
reg-tests/server/cli_add_server.vtc
reg-tests/http-rules/map_redirect.vtc

### Expected Behavior

stable behaviour

### Steps to Reproduce the Behavior

1. run vtc either on graviton2 or 4xOCPU Ampere A1

### What is your configuration?

```haproxy
CPU: Ampere A1 on Oracle Cloud
4xOCPU

Oracle Linux 8

gcc9
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6-dev1-6fed59-29 2022/02/12 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.4.17-2136.302.7.2.1.el8uek.aarch64 #2 SMP Tue Jan 18 12:11:39 PST 2022 aarch64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = /opt/rh/gcc-toolset-9/root/usr/bin/gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment -Werror
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_OBSOLETE_LINKER=1
  DEBUG   = -DDEBUG_STRICT=1

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD +OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with gcc compiler version 9.2.1 20191120 (Red Hat 9.2.1-2)
Encrypted password support via crypt(3): yes
Built with PCRE version : 8.42 2018-03-20
Running on PCRE version : 8.42 2018-03-20
PCRE library supports JIT : yes
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.
Built with OpenSSL version : OpenSSL 1.1.1k  FIPS 25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  FIPS 25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG

Available services : none

Available filters :
	[TRACE] trace
	[COMP] compression
	[FCGI] fcgi-app
	[CACHE] cache
	[SPOE] spoe
```

**Comment by wtarreau** [comment]
Interesting. Did you try on a different platform to get comparable numbers ? (e.g x86) ? I don't think the failure rate should depend on the platform, however it should depend on the OS latency. It *might* be possible that you don't have the CPUs 100% of the time and that such issues are caused by timeouts striking before some operations have the time to complete. If so we could possibly slightly increase timeouts where relevant. But it might also be possible that these tests trigger real bugs. Some VTC involving the CLI might possibly be a bit racy, due to limitations in our ability to sequence certain tests, and maybe these ones need to be improved (or fixed if at all possible).

**Comment by wtarreau** [comment]
There are definitely race conditions in such tests. For example today I saw "dynamic_server_ssl" fail for the first time. I looked at the test, at first glance it's valid, but nothing prevents a "delete server" operation from being started before the end of the previous request to be done, meaning that for a small time lapse there's randomly either 0 or 1 connection to the server, making the delete operation reject it. This tiny change reliably addresses it:
```
diff --git a/reg-tests/ssl/dynamic_server_ssl.vtc b/reg-tests/ssl/dynamic_server_ssl.vtc
index 0e6ecb5ab..bb052a56e 100644
--- a/reg-tests/ssl/dynamic_server_ssl.vtc
+++ b/reg-tests/ssl/dynamic_server_ssl.vtc
@@ -68,6 +68,13 @@ haproxy h1 -cli {
        send "disable server test/s3"
        expect ~ ".*"
 
+       send "shutdown sessions server test/s1"
+       expect ~ ".*"
+       send "shutdown sessions server test/s2"
+       expect ~ ".*"
+       send "shutdown sessions server test/s3"
+       expect ~ ".*"
+
        # valid command
        send "experimental-mode on; del server test/s1"
        expect ~ "Server deleted."
```

I just found better which also works fine (I'll merge it):
```
@@ -23,6 +23,7 @@ haproxy h1 -conf {
 
        defaults
                mode http
+               option httpclose
                timeout connect "${HAPROXY_TEST_TIMEOUT-5s}"
                timeout client  "${HAPROXY_TEST_TIMEOUT-5s}"
                timeout server  "${HAPROXY_TEST_TIMEOUT-5s}"
```

This will make sure we've already flushed the server connection when the client receives the end.

I've looked at cli_set_ssl.vtc and can't spot anything obvious, maybe its logs could be useful then.

**Comment by wtarreau** [comment]
I'm seeing that one fail ~1/200 times: tls_basic_sync_wo_stkt_backend.vtc
I don't understand why it works in fact, because at the end it does this:
```
haproxy h3 -cli {
    send "show table peers/stkt"
    expect ~ "# table: peers/stkt, type: string, size:1048[0-9]{4}, used:4(\n0x[0-9a-f]*: key=/c[1-4]_client use=0 exp=0 server_id=0 gpc0=1 conn_rate\\(50000\\)=1 conn_cur=0){4}"
}

haproxy h4 -cli {
    send "show table peers/stkt"
    expect ~ "# table: peers/stkt, type: string, size:1048[0-9]{4}, used:0\n"
}
```
It configures 4 instances exactly similarly with the same listening socket and expects that the first 3 ones are synchronized but that the 4th one is not. Hmmm no they're slightly different but that's not indicated in the configs. Some use SSL, others not. Some are declared in the peers section, others not. Some track TCP others HTTP. The reality here is that 1 over 200 times the 4th instance had the time to synchronize and also has its 4 entries properly learned. 
Apparently the real issue is the "delay 2" which is too short to allow for peers to resync when starting (there's probably one or two collisions, turn-around and new attempts). With 3 seconds it's always OK now. And the other delays are useless.

**Comment by wtarreau** [comment]
On my ARM board I'm seeing ssl_dh.vtc fail 100% of the time with a 503 instead of the expected 200:
```
***  h3    debug|<134>Feb 16 10:34:14 haproxy[1341185]: 127.0.0.1:14810 [16/Feb/2022:10:34:13.861] clear-ls
t clear-lst/s1 0/0/-1/-1/+426 503 +217 - - sC-- 2/1/0/0/0 0/0 "GET / HTTP/1.1" 0/-/-/-/0 -/-/-
***  h3    debug|<134>Feb 16 10:34:14 haproxy[1341185]: unix:1 [16/Feb/2022:10:34:14.182] ssl-dhfile-lst~ s
sl-dhfile-lst/<NOSRV> -1/-1/-1/-1/+105 400 +0 - - CR-- 2/1/0/0/0 0/0 "<BADREQ>" 0/0000000000000000/0/0/0 -/
TLSv1.2/DHE-RSA-AES256-GCM-SHA384
**** dT    0.918
**** c3    rxhdr|HTTP/1.1 503 Service Unavailable\r
**** c3    rxhdr|content-length: 107\r
**** c3    rxhdr|cache-control: no-cache\r
**** c3    rxhdr|content-type: text/html\r
**** c3    rxhdr|\r
**** c3    rxhdrlen = 107
**** c3    http[ 0] |HTTP/1.1
**** c3    http[ 1] |503
**** c3    http[ 2] |Service Unavailable
**** c3    http[ 3] |content-length: 107
**** c3    http[ 4] |cache-control: no-cache
**** c3    http[ 5] |content-type: text/html
**** c3    c-l|<html><body><h1>503 Service Unavailable</h1>
**** c3    c-l|No server is available to handle this request.
**** c3    c-l|</body></html>
**** c3    bodylen = 107
**   c3    === expect resp.status == 200
---- c3    EXPECT resp.status (503) == "200" failed
```
Maybe it doesn't accept the cipher in TLSv1.2, I don't know. This machine uses this:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.3 LTS
Release:        20.04
Codename:       focal
$ openssl version
OpenSSL 1.1.1f  31 Mar 2020
```

CCing @rlebreton in case he has any idea.

**Comment by wtarreau** [comment]
I also seem to notice that running with too large an nproc tends to increase the failure rate (even on x86). Probably that in the CI we should limit this (I think we're using nproc but we could also enforce it to 1) to avoid triggering false positives.

**Comment by wtarreau** [comment]
At least http_errors runs fine even for 1000 tests at -j16 here.

**Comment by wtarreau** [comment]
cli_set_ssl never fails for me (1000 times, -j16).
However cli_add_server fails 4/1000:
```
**** c3    rxhdrlen = 103
**** c3    http[ 0] |HTTP/1.1
**** c3    http[ 1] |504
**** c3    http[ 2] |Gateway Time-out
**** c3    http[ 3] |content-length: 92
**** c3    http[ 4] |cache-control: no-cache
**** c3    http[ 5] |content-type: text/html
**** c3    c-l|<html><body><h1>504 Gateway Time-out</h1>
**** c3    c-l|The server didn't respond in time.
**** c3    c-l|</body></html>
**** c3    bodylen = 92
**   c3    === expect resp.status == 200
---- c3    EXPECT resp.status (504) == "200" failed
```
Surprising, especially for the 504 which seems to indicate there was the start of a response.

I think I understand, it's randomly using the VTC's "server s1" as an H2 server, but this one is not configured for this. When it works it's probably thanks to retries. And maybe that the server keeps an incomplete request due to the H2 preface it receives. Strangely, changing it to "balance first" to always use server s1 doesn't fix it. I'll see if disabling the s2 server addresses it.

**Comment by chipitsine** [comment]
I did not use `-j` in above table. I use the following script

```
#!/bin/bash

export   FLAGS="USE_OPENSSL=1 USE_PCRE=1 USE_PCRE_JIT=1 USE_SYSTEMD=1 USE_ZLIB=1 USE_OBSOLETE_LINKER=1"
export   TMPDIR=/tmp

make clean
make ERR=1 TARGET=linux-glibc CC=/opt/rh/gcc-toolset-9/root/usr/bin/gcc DEBUG=-DDEBUG_STRICT=1 $FLAGS
fail=0

rm -rf /tmp/haregtests-*

for i in {1..1000}
do
   make reg-tests VTEST_PROGRAM=../vtest/vtest -- reg-tests/server/cli_set_ssl.vtc

   r=$?
   if [ $r -ne 0 ]; then
       fail=$((fail+1))
   fi

done

echo '================'
echo $fail
echo '==============='
```



I ran tests on Microsoft Azure today (on different VM size), will provide table tomorrow

**Comment by wtarreau** [comment]
Then that's strange. It could indicate remaining races in these checks, where some operations can fire and complete on a concurrent thread while they're being processed on the CLI.

Note, instead of the for() loop above you can simply pass `--vtestparams -n1000` to your `make` command line, and vtest will run the test 1000 times. That's way faster and will require less of your patience to get the result ;-)

**Comment by chipitsine** [comment]
we observed travis-ci/graviton2 failures several times. same tests ran on Github Actions (which I believe ran on Microsoft Azure, usually do not fail)

**Comment by chipitsine** [comment]
### Microsoft Azure x86_64

haproxy version:

```
commit d544d33e10868963407212660466aa552d5888e6 (HEAD -> master, origin/master, origin/HEAD)
Author: Remi Tricot-Le Breton <rlebreton@haproxy.com>
Date:   Fri Feb 4 14:24:15 2022 +0100
```

haproxy -vv

```
HAProxy version 2.6-dev1-d544d3-88 2022/02/15 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 4.18.0-193.28.1.el8_2.x86_64 #1 SMP Thu Oct 22 00:20:22 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = /opt/rh/gcc-toolset-9/root/usr/bin/gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment -Werror
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_OBSOLETE_LINKER=1
  DEBUG   = -DDEBUG_STRICT=1

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD +OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with gcc compiler version 9.2.1 20191120 (Red Hat 9.2.1-2)
Encrypted password support via crypt(3): yes
Built with PCRE version : 8.42 2018-03-20
Running on PCRE version : 8.42 2018-03-20
PCRE library supports JIT : yes
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.
Built with OpenSSL version : OpenSSL 1.1.1k  FIPS 25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  FIPS 25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG

Available services : none

Available filters :
        [TRACE] trace
        [COMP] compression
        [FCGI] fcgi-app
        [CACHE] cache
        [SPOE] spoe
```

CentOS  8.2.2004

test used: `reg-tests/server/cli_set_ssl.vtc`

| Size| number of failures (per 1000 execution) |
|--|--|
| B1ms| 0 |
| B2ms| 0 |
| B4ms| 0 |
| B16ms| 0 |
| F16s| 0 |
| D16s_v3| 0 |

**Comment by rlebreton** [comment]
@wtarreau Concerning the ssl_dh.vtc test, I don't see what could go wrong on ARM specifically yet. The SSL part of the connection seems to work fine and we fail when trying to talk to the most basic HTTP server there is. And since we're failing on the 'c3' client, the vtc server should only have received two requests when it is configured to manage 8 of them. Do you still have the full log somewhere ?

**Comment by wtarreau** [comment]
I just found! That was just a PEBKAC. I copied my "make-reg-tests-all" scripts which passes the environment variables to the machine and used it to start the test. When run manually it works. The difference is that for my PC it sets the test timeout to 400ms, and this is visibly not enough for large keys on the ARM board ;-)

Once this variable is removed, all tests now pass, both on a quad-A72 (armv8.0) and on a quad-A55 (armv8.2).
I'll send you private instructions to connect to the local test machine.
